device-connectors/src/testflinger_device_connectors/fw_devices/HPE/HPE.py

```diff
@@ -0,0 +1,456 @@
+                    json_index[spp] = json.load(json_index_file)
+        except IOError:
+            msg = f"Unable to open cached copy of index {spp}/{INDEX_FILE}"
+            raise RuntimeError(msg)
+        self.fwrepo_index = dict(sorted(json_index.items(), reverse=True))
+
+    def get_fw_info(self):
+        """
+        Get current firmware version of all updatable devices on HPE machine,
+        and print out devices with upgradable/downgradable versions
+        """
+        fw_inventory = self._rawget_firmware_inventory()
+        self._get_hpe_fw_repo()
+        server_fw_info = []
+        update_prio = [
+            "system rom",
+            "server platform services",
+            "innovation engine",
+            "ilo",
+        ]
+        for fw in fw_inventory:
+            fwpkg_data = {}
+            update_order = 5
+            if fw["Updateable"]:
+                update_order = 4
+                if fw["Oem"]["Hpe"].get("Targets") is not None:
+                    fwpkg_data = self._repo_search(
+                        fw["Oem"]["Hpe"].get("DeviceClass"),
+                        fw["Oem"]["Hpe"]["Targets"],
+                    )
+                for x in update_prio:
+                    if x in fw["Name"].lower():
+                        update_order = update_prio.index(x)
+            server_fw_info.append(
+                dict(
+                    {
+                        "Firmware Name": fw["Name"],
+                        "Firmware Version": fw["Version"],
+                        "Location": fw["Oem"]["Hpe"]["DeviceContext"],
+                        "Fwpkg Available": fwpkg_data,
+                        "Update Order": update_order,
+                    }
+                )
+            )
+        self.fw_info = sorted(server_fw_info, key=lambda x: x["Update Order"])
+        logmsg(logging.INFO, f"firmware on HPE machine:\n{self.fw_info}")
+
+    def _purify_ver(self, ver_string: str) -> str:
+        """
+        Extract pure version number and unify the format
+
+        :param ver_string: version from iLO or fwrepo.json
+        :return:           purified numeric version
+        """
+        ver_num = re.sub(
+            r"\.|\)|\(|\/|-|_",
+            " ",
+            re.match(r"([^a-zA-Z]*)", ver_string.split("v")[-1])
+            .group(1)
+            .replace(" ", ""),
+        ).strip()
+        return ".".join(map(str, list(map(int, ver_num.split(" ")))))
+
+    def _flash_fwpkg(self, spp: str) -> bool:
+        """
+        Flash FW via ilorest and return `True` if a machine reboot is needed
+
+        :param spp: HPE SPP released date = repo folder name
+        :return:    `True` if reboot is needed, `False` otherwise
+        """
+        if spp not in str(self.fw_info):
+            raise RuntimeError(
+                f"SPP {spp} is not available in HPE FW repository."
+                " Please check if it's a valid SPP."
+            )
+        logmsg(
```

Comment:
I have a WIP branch that removes logmsg(). I would definitely recommend replacing all new calls you have here to logmsg to just do the normal logger.{loglevel}(...)

Reply:
Removed logmsg().